File: generate.py
#!/usr/local/bin/python3
import os, datetime, re, json, requests
from TwitterAPI import TwitterAPI

def is_leap_year(y):
    return y % 4 == 0 and (y % 100 != 0 or y % 400 == 0)

def prepend_zero(n):
    return "0" + str(n) if n < 10 else n

def day_suffix(n):
    return "th" if 4 <= n <= 20 or 24 <= n <= 30 else ["st", "nd", "rd"][n % 10 - 1]

month_days  = [31, 0, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
month_names = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"]
sub_money = { 'Prime': 2.5, '1000': 2.5, '2000': 6.66, '3000': 17.5 }

sub_re = re.compile(r'^\[(\d){2}:(\d){2}\] <twitchnotify> \S+ (.*)$')
new_sub_re = re.compile(r'.*\\ssubscribed\\s')
cheer_re = re.compile(r'[?<=\s](?:kappa|kreygasm|swiftrage|cheer)(\d+)(?!\S)')

c_date = datetime.datetime.now()
last_month = month_names[11 if c_date.month == 1 else c_date.month - 2]

users = {}
paypigs = {}
last_month_paypigs = {}
data = {}
for year in range(2016, c_date.year + 1):
    data[year] = {}
    for month in range(0, 12):
        month_name = month_names[month]
        if os.path.exists("logs/{}/{}".format(year, month_name)):
            data[year][month_name] = {}
            for day in range(1, (29 if is_leap_year(year) else 28 if month == 1 else month_days[month]) + 1):
                path = "logs/{}/{}/{}-{}-{}.txt".format(year, month_name, year, prepend_zero(month + 1), prepend_zero(day))
                if os.path.exists(path):
                    data[year][month_name][day] = { 'permabans': 0, 'bans': 0, 'bits': 0, 'subs': 0 }
                    with open(path, 'r') as fh:
                        if year > 2016:
                            for line in fh:
                                parts = line.split(' #darksydephil :', 1)
                                info  = dict(x.split('=') for x in parts[0].split(';'))
                                msg   = parts[1][:-1] if len(parts) == 2 else ""
                                if 'user-id' in info:
                                    if not info['user-id'] in users:
                                        users[info['user-id']] = []
                                    if 'display-name' in info and info['display-name'] not in users[info['user-id']] and info['display-name'] != '':
                                        users[info['user-id']].append(info['display-name'])
                                    else:
                                        name = info['user-type'].split('!')[0][2:]
                                        if name not in users[info['user-id']]:
                                            users[info['user-id']].append(name)

                                    if 'bits' in info:
                                        bits = int(info['bits'])
                                        data[year][month_name][day]['bits'] += bits
                                        if not info['user-id'] in paypigs:
                                            paypigs[info['user-id']] = 0
                                        paypigs[info['user-id']] += bits
                                        if year == c_date.year and month_name == last_month:
                                            if not info['user-id'] in last_month_paypigs:
                                                last_month_paypigs[info['user-id']] = 0
                                            last_month_paypigs[info['user-id']] += bits
                                    elif 'msg-param-sub-plan' in info:
                                        data[year][month_name][day]['subs'] += sub_money[info['msg-param-sub-plan']]
                                    elif 'subscriber' in info and info['subscriber'] == '1' and 'system-msg' in info and new_sub_re.match(info['system-msg']):
                                        data[year][month_name][day]['subs'] += 2.5;

                                else:
                                    data[year][month_name][day]['bans' if '@ban-duration' in info else 'permabans'] += 1
                        else:
                            for line in fh:
                                m = sub_re.match(line)
                                if m:
                                    data[year][month_name][day]['subs'] += 2.5
                                else:
                                    data[year][month_name][day]['bits'] += sum([int(x) for x in cheer_re.findall(line)])

paypigs_out = {}
last_month_paypigs_out = {}

paypigs_out = { users[v[0]][0]: v[1] for v in sorted(paypigs.items(), key=lambda k: k[1])[-30:] }
last_month_paypigs_out = { users[v[0]][0]: v[1] for v in sorted(last_month_paypigs.items(), key=lambda k: k[1])[-30:] }

patreon_data = None
for line in requests.get("https://graphtreon.com/creator/darksydephil").text.split("\n"):
    if line[2:14] == "var dataJson":
        patreon_data = line[2:]
        break

youtube_data = None
for line in requests.get("https://socialblade.com/youtube/user/dspgaming/monthly").text.split("\n"):
    if line[2:20] == 'Date,Average Views':
        youtube_data = dict(y.split(',') for y in [x[1:-3] for x in line[26:-4].split(" + ")])
        break

with open("www/data.js", 'w') as fh:
    fh.write("var data = '{}';\n{}\nvar yt_data = '{}';\nvar paypigs = '{}';\nvar last_paypigs = '{}';\nvar last_patreon = {};\nvar last_update = '{}';".format(json.dumps(data), patreon_data, json.dumps(youtube_data), json.dumps(paypigs_out), json.dumps(last_month_paypigs_out), int(re.findall(r'"pledge_sum": (\d+),', requests.get("https://www.patreon.com/darksydephil").text)[0]) / 100, datetime.datetime.now().strftime("%s")).replace('\\s', ''))

year = c_date.year
month = c_date.month
day = c_date.day

if day == 1:
    month = month - 1 if month > 1 else 12
    if month == 12:
        year -= 1
    day = (29 if is_leap_year(year) else 28 if month == 2 else month_days[month - 1])
else:
    day -= 1
month = month_names[month - 1]

config = [x[:-1] for x in open('twitter_bot.conf', 'r').readlines() if len(x) > 1]
print(TwitterAPI(config[0], config[1], config[2], config[3]).request('statuses/update', { 'status': "#DSP #TheSnortReport for {}{} {}, {}: Cheers: ${}, Subs: ${}".format(day, day_suffix(day), month, year, data[year][month][day]['bits'] / 100, data[year][month][day]['subs']) }).status_code)

File: generate.rb
#!/usr/bin/env ruby
require 'json'
require 'twitter'
require 'net/http'

month_names = %w[January February March April May June July August September October November December]
month_days  = [31, 0, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
sub_money   = { 'Prime' => 2.5,
                '1000'  => 2.5,
                '2000'  => 6.66,
                '3000'  => 17.5 }

def is_leap(y)
  y % 4 == 0 && (y % 100 != 0 || y % 400 == 0)
end

def date_suffix(n)
  if (11..13).cover? n.to_i.abs % 100
    'th'
  else
    case n.to_i.abs % 10
    when 1 then 'st'
    when 2 then 'nd'
    when 3 then 'rd'
    else        'th'
    end
  end
end

class String
  def is_i?
    /\A[-+]?\d+\z/ === self
  end
end

now = Time.now
last_month = month_names[(now.month == 12 ? 1 : now.month) - 1]
data = {}
users = {}
paypigs = {}
last_month_paypigs = {}
2016.upto now.year do |year|
  data[year] = {}
  0.upto 11 do |month_n|
    month = month_names[month_n]
    data[year][month] = {}
    1.upto (month_n == 1 ? (is_leap(year) ? 29 : 28) : month_days[month_n]) do |day|
      fh_p = "./logs/#{year}/#{month}/#{year}-#{(month_n + 1).to_s.rjust(2, '0')}-#{day.to_s.rjust(2, '0')}.txt"
      next unless File.exist? fh_p
      data[year][month][day] = { 'permabans' => 0,
                                 'bans'      => 0,
                                 'bits'      => 0,
                                 'subs'      => 0,
                                 'tips'      => 0 }

      File.open fh_p, 'r' do |fh|
        while line = fh.gets
          if year == 2016
            if line =~ /^\[(\d){2}:(\d){2}\] <twitchnotify> \S+ (.*)$/
              data[year][month][day]['subs'] += 2.5
            else
              m = line.scan(/[?<=\s](?:kappa|kreygasm|swiftrage|cheer)(\d+)(?!\S)/)
              data[year][month][day]['bits'] += m.flatten(1).map(&:to_i).inject(0, :+) unless m.empty?
            end
          else
            next unless line[0] == '@'

            /^@(?<params>[a-zA-Z-]+=\S*)\s?:((?<user>\S+!\S+@\S+)\.)?tmi\.twitch\.tv (?<type>\S+) #\S+( :(?<msg>.*))?$/ =~ line
            unless $~.nil?
              params = params.split(';').reject { |a| a[-1] == '=' }.map { |b| b.split '=' }.to_h
              user   = user.split('!')[0] unless user.nil?

              if params.key? 'user-id'
                users[params['user-id']] = [] unless users.key? 'user-id'
                if params.key?('display-name') && !users[params['user-id']].include?(params['display-name'])
                  users[params['user-id']].push params['display-name']
                else
                  unless user.nil?
                    name = user.split('!')[0]
                    users[params['user-id']].push name unless users[params['user-id']].include? name
                  end
                end

                if (year == 2017 && month_n >= 10 || year > 2017) && (params['user-id'] == '19264788' && params['display-name'] == 'Nightbot')
                  /(.*) just tipped \$([-+]?[0-9]*\.?[0-9]+)/ =~ msg
                  unless $~.nil?
                    tips = $2.to_f
                    data[year][month][day]['tips'] += tips
                    user_id = nil
                    unless $1 == "Anonymous"
                      users.each do |k, v|
                        if v.any? { |s| s.casecmp($1) == 0 }
                          user_id = k
                          break
                        end
                      end
                    end
                    user_id = $1 if user_id.nil?
                    paypigs[user_id]  = 0 unless paypigs.key? user_id
                    paypigs[user_id] += tips
                    if year == now.year && month == last_month
                      last_month_paypigs[user_id]  = 0 unless last_month_paypigs.key? user_id
                      last_month_paypigs[user_id] += tips
                    end
                  end
                elsif params.key? 'bits'
                  bits = params['bits'].to_i
                  data[year][month][day]['bits'] += bits
                  paypigs[params['user-id']]  = 0 unless paypigs.key? params['user-id']
                  paypigs[params['user-id']] += bits
                  if year == now.year && month == last_month
                    last_month_paypigs[params['user-id']]  = 0 unless last_month_paypigs.key? params['user-id']
                    last_month_paypigs[params['user-id']] += bits
                  end
                elsif params.key? 'msg-param-sub-plan'
                  data[year][month][day]['subs'] += sub_money[params['msg-param-sub-plan']]
                elsif params.key?('subscriber') && params['subscriber'] == '1' && params.key?('system-msg') && params['system-msg'] =~ /.*\\ssubscribed\\s/
                  data[year][month][day]['subs'] += 2.5
                end
              else
                data[year][month][day][(params.key?('ban-duration') ? 'bans' : 'permabans')] += 1
              end
            end
          end
        end
      end
    end
    data[year].delete(month) if data[year][month].empty?
  end
end

paypigs = paypigs.sort_by { |_k, v| v }.reverse[0..30].map { |v| [(v[0].is_i? ? users[v[0]][0] : v[0]), v[1]] }.to_h.to_json
last_month_paypigs = last_month_paypigs.sort_by { |_k, v| v }.reverse[0..30].map { |v| [(v[0].is_i? ? users[v[0]][0] : v[0]), v[1]] }.to_h.to_json

pt = /var dataJson = '(.*)'/.match(Net::HTTP.get(URI('https://graphtreon.com/creator/darksydephil')))[1]
yt = /"Date,Average Views\\n" \+ (.*) {/.match(Net::HTTP.get(URI('https://socialblade.com/youtube/user/dspgaming/monthly')))[1].scan(/(\d+-\d+-\d+,\d+)/).map { |x| x[0].split ',' }.to_h.to_json
lp = Net::HTTP.get(URI('https://www.patreon.com/darksydephil')).scan(/pledge_sum": (\d+)/)[0][0]

puts "var data = '#{data.to_json}';\nvar dataJson = '#{pt}'\nvar yt_data = '#{yt}';\nvar paypigs = '#{paypigs}';\nvar last_paypigs = '#{last_month_paypigs}';\nvar last_patreon = #{lp.to_i / 100};\nvar last_update = #{now.to_i};"

exit 0 # TODO: Remove once finished testing

year  = now.year
month = now.month
day   = now.day

if day == 1
  month = (month > 1 ? month - 1 : 12)
  year -= 1 if month == 12
  day   = (month == 2 ? (is_leap(year) ? 29 : 28) : month_days[month - 1])
else
  day -= 1
end
month = month_names[month - 1]

a, b, c, d = File.open('twitter.conf').readlines.map(&:rstrip)
client = Twitter::Streaming::Client.new do |config|
  config.consumer_key        = a
  config.consumer_secret     = b
  config.access_token        = c
  config.access_token_secret = d
end
client.update("#DSP #TheSnortReport for #{day}#{day_suffix day} #{month}, #{year}: Cheers: $#{data[year][month][day]['bits'] / 100}, Subs: $#{data[year][month][day]['subs']}")
